executor/server.go

```diff
@@ -260,29 +282,96 @@ func (s *Server) keepalive(ctx context.Context) error {
 		},
 	})
 	if err != nil {
-		return err
+		return nil, err
+	}
+	session, err := s.createSession(ctx, etcdCli)
+	if err != nil {
+		return nil, err
 	}
+
+	// initialize a new service discovery, if old discovery exists, clones its
+	// snapshot to the new one.
+	old := s.discovery
+	s.discovery = srvdiscovery.NewEtcdSrvDiscovery(
+		etcdCli, adapter.ExecutorInfoKeyAdapter, defaultDiscoverTicker)
+	if old != nil {
+		s.discovery.CopySnapshot(old.SnapshotClone())
+	}
+
+	return session, nil
+}
+
+func (s *Server) createSession(ctx context.Context, etcdCli *clientv3.Client) (metaStoreSession, error) {
 	session, err := concurrency.NewSession(etcdCli, concurrency.WithTTL(s.cfg.SessionTTL))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	// TODO: we share system metastore with service discovery etcd, in the future
 	// we could separate them
 	s.metastore = metadata.NewMetaEtcd(etcdCli)
 	value, err := s.info.ToJSON()
 	if err != nil {
-		return err
+		return nil, err
 	}
 	_, err = s.metastore.Put(ctx, s.info.EtcdKey(), value, clientv3.WithLease(session.Lease()))
+	if err != nil {
+		return nil, err
+	}
+	return session, nil
+}
+
+func (s *Server) discoveryKeepalive(ctx context.Context) error {
+	var (
+		session metaStoreSession
+		err     error
+	)
+	session, err = s.discoveryConnector(ctx)
+	if err != nil {
+		return err
+	}
+	executors, err := s.discovery.Snapshot(ctx)
 	if err != nil {
 		return err
 	}
+	for uuid, exec := range executors {
+		if uuid == string(s.info.ID) {
+			continue
```

Comment:
I think we should not skip adding itself to the router.

Reply:
addressed